// recorder-audio-worklet/build/es5/bundle.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('fast-unique-numbers'), require('subscribable-things'), require('worker-factory'), require('@babel/runtime/helpers/asyncToGenerator'), require('@babel/runtime/regenerator'), require('@babel/runtime/helpers/defineProperty')) :
    typeof define === 'function' && define.amd ? define(['exports', 'fast-unique-numbers', 'subscribable-things', 'worker-factory', '@babel/runtime/helpers/asyncToGenerator', '@babel/runtime/regenerator', '@babel/runtime/helpers/defineProperty'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.recorderAudioWorklet = {}, global.fastUniqueNumbers, global.subscribableThings, global.workerFactory, global._asyncToGenerator, global._regeneratorRuntime, global._defineProperty));
})(this, (function (exports, fastUniqueNumbers, subscribableThings, workerFactory, _asyncToGenerator, _regeneratorRuntime, _defineProperty) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
    var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);
    var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

    var createAddRecorderAudioWorkletModule = function createAddRecorderAudioWorkletModule(blobConstructor, urlConstructor, worklet) {
      return /*#__PURE__*/function () {
        var _ref = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee(addAudioWorkletModule) {
          var blob, url;
          return _regeneratorRuntime__default["default"].wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  blob = new blobConstructor([worklet], {
                    type: 'application/javascript; charset=utf-8'
                  });
                  url = urlConstructor.createObjectURL(blob);
                  _context.prev = 2;
                  _context.next = 5;
                  return addAudioWorkletModule(url);

                case 5:
                  _context.prev = 5;
                  urlConstructor.revokeObjectURL(url);
                  return _context.finish(5);

                case 8:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, null, [[2,, 5, 8]]);
        }));

        return function (_x) {
          return _ref.apply(this, arguments);
        };
      }();
    };

    var createListener = function createListener(ongoingRequests) {
      return function (_ref) {
        var message = _ref.data;
        var id = message.id;

        if (id !== null) {
          var ongoingRequest = ongoingRequests.get(id);

          if (ongoingRequest !== undefined) {
            var reject = ongoingRequest.reject,
                resolve = ongoingRequest.resolve;
            ongoingRequests["delete"](id);

            if (message.error === undefined) {
              resolve(message.result);
            } else {
              reject(new Error(message.error.message));
            }
          }
        }
      };
    };

    function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

    function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys$1(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys$1(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

    var createPostMessageFactory = function createPostMessageFactory(generateUniqueNumber) {
      return function (ongoingRequests, port) {
        return function (message) {
          var transferables = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
          return new Promise(function (resolve, reject) {
            var id = generateUniqueNumber(ongoingRequests);
            ongoingRequests.set(id, {
              reject: reject,
              resolve: resolve
            });
            port.postMessage(_objectSpread$1({
              id: id
            }, message), transferables);
          });
        };
      };
    };

    function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

    function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

    var createRecorderAudioWorkletNodeFactory = function createRecorderAudioWorkletNodeFactory(createListener, createPostMessage, on, validateState) {
      return function (audioWorkletNodeConstructor, context) {
        var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
        var audioWorkletNode = new audioWorkletNodeConstructor(context, 'recorder-audio-worklet-processor', _objectSpread(_objectSpread({}, options), {}, {
          channelCountMode: 'explicit',
          numberOfInputs: 1,
          numberOfOutputs: 0
        }));
        var ongoingRequests = new Map();
        var postMessage = createPostMessage(ongoingRequests, audioWorkletNode.port);
        var unsubscribe = on(audioWorkletNode.port, 'message')(createListener(ongoingRequests));
        audioWorkletNode.port.start();
        var state = 'inactive';
        Object.defineProperties(audioWorkletNode, {
          pause: {
            get: function get() {
              return /*#__PURE__*/_asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee() {
                return _regeneratorRuntime__default["default"].wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        validateState(['recording'], state);
                        state = 'paused';
                        return _context.abrupt("return", postMessage({
                          method: 'pause'
                        }));

                      case 3:
                      case "end":
                        return _context.stop();
                    }
                  }
                }, _callee);
              }));
            }
          },
          port: {
            get: function get() {
              throw new Error("The port of a RecorderAudioWorkletNode can't be accessed.");
            }
          },
          record: {
            get: function get() {
              return /*#__PURE__*/function () {
                var _ref2 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee2(encoderPort) {
                  return _regeneratorRuntime__default["default"].wrap(function _callee2$(_context2) {
                    while (1) {
                      switch (_context2.prev = _context2.next) {
                        case 0:
                          validateState(['inactive'], state);
                          state = 'recording';
                          return _context2.abrupt("return", postMessage({
                            method: 'record',
                            params: {
                              encoderPort: encoderPort
                            }
                          }, [encoderPort]));

                        case 3:
                        case "end":
                          return _context2.stop();
                      }
                    }
                  }, _callee2);
                }));

                return function (_x) {
                  return _ref2.apply(this, arguments);
                };
              }();
            }
          },
          resume: {
            get: function get() {
              return /*#__PURE__*/_asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee3() {
                return _regeneratorRuntime__default["default"].wrap(function _callee3$(_context3) {
                  while (1) {
                    switch (_context3.prev = _context3.next) {
                      case 0:
                        validateState(['paused'], state);
                        state = 'recording';
                        return _context3.abrupt("return", postMessage({
                          method: 'resume'
                        }));

                      case 3:
                      case "end":
                        return _context3.stop();
                    }
                  }
                }, _callee3);
              }));
            }
          },
          stop: {
            get: function get() {
              return /*#__PURE__*/_asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee4() {
                return _regeneratorRuntime__default["default"].wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        validateState(['paused', 'recording'], state);
                        state = 'stopped';
                        _context4.prev = 2;
                        _context4.next = 5;
                        return postMessage({
                          method: 'stop'
                        });

                      case 5:
                        _context4.prev = 5;
                        unsubscribe();
                        return _context4.finish(5);

                      case 8:
                      case "end":
                        return _context4.stop();
                    }
                  }
                }, _callee4, null, [[2,, 5, 8]]);
              }));
            }
          }
        });
        return audioWorkletNode;
      };
    };

    var validateState = function validateState(expectedStates, currentState) {
      if (!expectedStates.includes(currentState)) {
        throw new Error("Expected the state to be ".concat(expectedStates.map(function (expectedState) {
          return "\"".concat(expectedState, "\"");
        }).join(' or '), " but it was \"").concat(currentState, "\"."));
      }
    };

    // This is the minified and stringified code of the recorder-audio-worklet-processor package.
    var worklet = "(()=>{var e={228:e=>{e.exports=function(e,t){(null==t||t>e.length)&&(t=e.length);for(var r=0,o=new Array(t);r<t;r++)o[r]=e[r];return o},e.exports.__esModule=!0,e.exports.default=e.exports},858:e=>{e.exports=function(e){if(Array.isArray(e))return e},e.exports.__esModule=!0,e.exports.default=e.exports},506:e=>{e.exports=function(e){if(void 0===e)throw new ReferenceError(\"this hasn't been initialised - super() hasn't been called\");return e},e.exports.__esModule=!0,e.exports.default=e.exports},575:e=>{e.exports=function(e,t){if(!(e instanceof t))throw new TypeError(\"Cannot call a class as a function\")},e.exports.__esModule=!0,e.exports.default=e.exports},100:(e,t,r)=>{var o=r(489),n=r(67);function s(t,r,u){return n()?(e.exports=s=Reflect.construct.bind(),e.exports.__esModule=!0,e.exports.default=e.exports):(e.exports=s=function(e,t,r){var n=[null];n.push.apply(n,t);var s=new(Function.bind.apply(e,n));return r&&o(s,r.prototype),s},e.exports.__esModule=!0,e.exports.default=e.exports),s.apply(null,arguments)}e.exports=s,e.exports.__esModule=!0,e.exports.default=e.exports},913:e=>{function t(e,t){for(var r=0;r<t.length;r++){var o=t[r];o.enumerable=o.enumerable||!1,o.configurable=!0,\"value\"in o&&(o.writable=!0),Object.defineProperty(e,o.key,o)}}e.exports=function(e,r,o){return r&&t(e.prototype,r),o&&t(e,o),Object.defineProperty(e,\"prototype\",{writable:!1}),e},e.exports.__esModule=!0,e.exports.default=e.exports},754:e=>{function t(r){return e.exports=t=Object.setPrototypeOf?Object.getPrototypeOf.bind():function(e){return e.__proto__||Object.getPrototypeOf(e)},e.exports.__esModule=!0,e.exports.default=e.exports,t(r)}e.exports=t,e.exports.__esModule=!0,e.exports.default=e.exports},205:(e,t,r)=>{var o=r(489);e.exports=function(e,t){if(\"function\"!=typeof t&&null!==t)throw new TypeError(\"Super expression must either be null or a function\");e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,writable:!0,configurable:!0}}),Object.defineProperty(e,\"prototype\",{writable:!1}),t&&o(e,t)},e.exports.__esModule=!0,e.exports.default=e.exports},430:e=>{e.exports=function(e){return-1!==Function.toString.call(e).indexOf(\"[native code]\")},e.exports.__esModule=!0,e.exports.default=e.exports},67:e=>{e.exports=function(){if(\"undefined\"==typeof Reflect||!Reflect.construct)return!1;if(Reflect.construct.sham)return!1;if(\"function\"==typeof Proxy)return!0;try{return Boolean.prototype.valueOf.call(Reflect.construct(Boolean,[],(function(){}))),!0}catch(e){return!1}},e.exports.__esModule=!0,e.exports.default=e.exports},884:e=>{e.exports=function(e,t){var r=null==e?null:\"undefined\"!=typeof Symbol&&e[Symbol.iterator]||e[\"@@iterator\"];if(null!=r){var o,n,s=[],u=!0,p=!1;try{for(r=r.call(e);!(u=(o=r.next()).done)&&(s.push(o.value),!t||s.length!==t);u=!0);}catch(e){p=!0,n=e}finally{try{u||null==r.return||r.return()}finally{if(p)throw n}}return s}},e.exports.__esModule=!0,e.exports.default=e.exports},521:e=>{e.exports=function(){throw new TypeError(\"Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\")},e.exports.__esModule=!0,e.exports.default=e.exports},585:(e,t,r)=>{var o=r(8).default,n=r(506);e.exports=function(e,t){if(t&&(\"object\"===o(t)||\"function\"==typeof t))return t;if(void 0!==t)throw new TypeError(\"Derived constructors may only return object or undefined\");return n(e)},e.exports.__esModule=!0,e.exports.default=e.exports},489:e=>{function t(r,o){return e.exports=t=Object.setPrototypeOf?Object.setPrototypeOf.bind():function(e,t){return e.__proto__=t,e},e.exports.__esModule=!0,e.exports.default=e.exports,t(r,o)}e.exports=t,e.exports.__esModule=!0,e.exports.default=e.exports},38:(e,t,r)=>{var o=r(858),n=r(884),s=r(379),u=r(521);e.exports=function(e,t){return o(e)||n(e,t)||s(e,t)||u()},e.exports.__esModule=!0,e.exports.default=e.exports},8:e=>{function t(r){return e.exports=t=\"function\"==typeof Symbol&&\"symbol\"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&\"function\"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?\"symbol\":typeof e},e.exports.__esModule=!0,e.exports.default=e.exports,t(r)}e.exports=t,e.exports.__esModule=!0,e.exports.default=e.exports},379:(e,t,r)=>{var o=r(228);e.exports=function(e,t){if(e){if(\"string\"==typeof e)return o(e,t);var r=Object.prototype.toString.call(e).slice(8,-1);return\"Object\"===r&&e.constructor&&(r=e.constructor.name),\"Map\"===r||\"Set\"===r?Array.from(e):\"Arguments\"===r||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)?o(e,t):void 0}},e.exports.__esModule=!0,e.exports.default=e.exports},957:(e,t,r)=>{var o=r(754),n=r(489),s=r(430),u=r(100);function p(t){var r=\"function\"==typeof Map?new Map:void 0;return e.exports=p=function(e){if(null===e||!s(e))return e;if(\"function\"!=typeof e)throw new TypeError(\"Super expression must either be null or a function\");if(void 0!==r){if(r.has(e))return r.get(e);r.set(e,t)}function t(){return u(e,arguments,o(this).constructor)}return t.prototype=Object.create(e.prototype,{constructor:{value:t,enumerable:!1,writable:!0,configurable:!0}}),n(t,e)},e.exports.__esModule=!0,e.exports.default=e.exports,p(t)}e.exports=p,e.exports.__esModule=!0,e.exports.default=e.exports},242:function(e,t,r){!function(e,t,r,o,n,s,u){\"use strict\";function p(e){return e&&\"object\"==typeof e&&\"default\"in e?e:{default:e}}var a=p(e),i=p(t),l=p(r),c=p(o),f=p(n),d=p(s);function x(e){var t=_();return function(){var r,o=d.default(e);if(t){var n=d.default(this).constructor;r=Reflect.construct(o,arguments,n)}else r=o.apply(this,arguments);return f.default(this,r)}}function _(){if(\"undefined\"==typeof Reflect||!Reflect.construct)return!1;if(Reflect.construct.sham)return!1;if(\"function\"==typeof Proxy)return!0;try{return Boolean.prototype.valueOf.call(Reflect.construct(Boolean,[],(function(){}))),!0}catch(e){return!1}}var y=function(e){c.default(r,e);var t=x(r);function r(){var e;return i.default(this,r),(e=t.call(this))._encoderPort=null,e._state=\"inactive\",e.port.onmessage=function(t){var r=t.data;\"pause\"===r.method?\"active\"===e._state||\"recording\"===e._state?(e._state=\"paused\",e._sendAcknowledgement(r.id)):e._sendUnexpectedStateError(r.id):\"record\"===r.method?\"inactive\"===e._state?(e._encoderPort=r.params.encoderPort,e._state=\"active\",e._sendAcknowledgement(r.id)):e._sendUnexpectedStateError(r.id):\"resume\"===r.method?\"paused\"===e._state?(e._state=\"active\",e._sendAcknowledgement(r.id)):e._sendUnexpectedStateError(r.id):\"stop\"===r.method?\"active\"!==e._state&&\"paused\"!==e._state&&\"recording\"!==e._state||null===e._encoderPort?e._sendUnexpectedStateError(r.id):(e._stop(e._encoderPort),e._sendAcknowledgement(r.id)):\"number\"==typeof r.id&&e.port.postMessage({error:{code:-32601,message:\"The requested method is not supported.\"},id:r.id})},e}return l.default(r,[{key:\"process\",value:function(e){var t=a.default(e,1)[0];if(\"inactive\"===this._state||\"paused\"===this._state)return!0;if(\"active\"===this._state){if(void 0===t)throw new Error(\"No channelData was received for the first input.\");if(0===t.length)return!0;this._state=\"recording\"}if(\"recording\"===this._state&&null!==this._encoderPort){if(void 0===t)throw new Error(\"No channelData was received for the first input.\");if(0!==t.length)return this._encoderPort.postMessage(t,t.map((function(e){return e.buffer}))),!0;this._stop(this._encoderPort)}return!1}},{key:\"_sendAcknowledgement\",value:function(e){this.port.postMessage({id:e,result:null})}},{key:\"_sendUnexpectedStateError\",value:function(e){this.port.postMessage({error:{code:-32603,message:\"The internal state does not allow to process the given message.\"},id:e})}},{key:\"_stop\",value:function(e){e.postMessage([]),e.close(),this._encoderPort=null,this._state=\"stopped\"}}]),r}(p(u).default(AudioWorkletProcessor));y.parameterDescriptors=[],registerProcessor(\"recorder-audio-worklet-processor\",y)}(r(38),r(575),r(913),r(205),r(585),r(754),r(957))}},t={};function r(o){var n=t[o];if(void 0!==n)return n.exports;var s=t[o]={exports:{}};return e[o].call(s.exports,s,s.exports,r),s.exports}r.n=e=>{var t=e&&e.__esModule?()=>e.default:()=>e;return r.d(t,{a:t}),t},r.d=(e,t)=>{for(var o in t)r.o(t,o)&&!r.o(e,o)&&Object.defineProperty(e,o,{enumerable:!0,get:t[o]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),(()=>{\"use strict\";r(242)})()})();"; // tslint:disable-line:max-line-length

    var addRecorderAudioWorkletModule = createAddRecorderAudioWorkletModule(Blob, URL, worklet);
    var createRecorderAudioWorkletNode = createRecorderAudioWorkletNodeFactory(createListener, createPostMessageFactory(fastUniqueNumbers.generateUniqueNumber), subscribableThings.on, validateState);

    Object.defineProperty(exports, 'isSupported', {
        enumerable: true,
        get: function () { return workerFactory.isSupported; }
    });
    exports.addRecorderAudioWorkletModule = addRecorderAudioWorkletModule;
    exports.createRecorderAudioWorkletNode = createRecorderAudioWorkletNode;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
